File: Anagrams.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# vim:fenc=utf-8
#
# Copyright © 2014 fuel <fuel@fuel-3330>
#
# Distributed under terms of the MIT license.

import collections

class Solution:
    # @param strs, a list of strings
    # @return a list of strings
    def anagrams(self, strs):
        D=collections.defaultdict(list)
        for s in strs:
            D[''.join(sorted(s))].append(s)
        res = []
        for los in D.values():
            if(len(los)>=2):
                res.extend(los)
        return res

print Solution().anagrams(["dog","cat","god","tac"])
print Solution().anagrams(["cat","rye","aye","dog", "god","cud","cat","old","fop","bra"])

File: Algorithms/Anagrams.cpp
/*
 * Anagrams.cpp
 * Copyright (C) 2014 fuel <fuel@fuel-3330>
 *
 * Distributed under terms of the MIT license.
 */

#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>
using namespace std;
class Solution {
public:
    vector<string> anagrams(vector<string> &strs) {
        unordered_map<string,vector<string> > D;
        for(auto &s : strs) {
            string soredS = s;
            sort(soredS.begin(),soredS.end());
            D[soredS].push_back(s);
        }
        vector<string> res;
        for (auto &p : D) {
            if(p.second.size()>=2)
                res.insert(res.end(),p.second.begin(),p.second.end());
        }
        return res;
    }
};
